### ISSUE TITLE
afterOn executing before On

### ISSUE THREAD

**Issue by trharoldsen** [JetBrains/spek#39]
Expected: on runs before afterOn
Actual: afterOn runs before on

Sample:
```
given("") {
    beforeOn { print("1") }
        on("") {
            print("2")
            it("") {
                print("3")
            }
    }
    afterOn { print("4") }
}
```

output:
1423

**Comment by hhariri** [comment]
Yep it's a known issue. With the restructuring hopefully this will be solved.

**Comment by jskierbi** [comment]
Anything new in this respect?

I understand that ```beforeOn``` and ```afterOn``` are called during adding test task to the queue, with queue itself being processed afterwards. Is there anything I can be helpful with? I really like Spek and want to add it to my project badly, but this issue prevents me from doing so

**Comment by hhariri** [comment]
Is there any possibility of working around it by just calling the code you need yourself in a function call explicitly? Currently because of how Spek is designed, without major changes it's not possible :(

**Comment by jskierbi** [comment]
Hmm I forked repo and started some code rework in this regard, now I've managed to change structure a bit so that given, on and it calls can be deferred after listing them (in master this is done via ```iterateOn```). One problem yet to overcome -> JUnits ParentRunner need to list all children tree before execution, which is unfortunately not possible before running given/on/it expressions (registers children), and this means all expressions are called prior to test execution... Will dig a bit more into how JUnit ParentRunner works and hopefully find a solution :)

Please tell me, are those major changes planned? I don't want to overlap the code - in hope that my changes can be later accepted as a PR :) (I'll ut them ofc!)

**Comment by hastebrot** [comment]
Hi Jakub,

I experimented a bit with ParentRunner a few months ago. Register new
children during test execution works fine.

My plan was to use graph transformations to rewrite Spek classes to graphs
that resemble the execution order and use these graphs for the ParentRunner
tree and test run output. This approach could help to understand and debug
the execution order better and allow custom DSLs, e.g. a Mocha.js-like
hierarchical spec definition.

One problem is, that you have to run parts of the tests in order to get
parts of the spec. The graph transformation approach would traverse the
spec, run a part of the specification, add new nodes to the graph,
transform it to be in execution order and resume the execution.

Maybe this approach (graph transformations) is a bit overkill. AFAIK other
test frameworks don't use this approach. But it is quite popular in the
compiler-world, I saw Oracle's GraalVM (which will be part of JDK 9) uses
it.

**Comment by hhariri** [comment]
@jskierbi Currently nothing planned, so your PR is more than welcome :)

**Comment by jskierbi** [comment]
Hi. Was exploring ParentRunner today, indeed there's an option to modify description (children etc) during test execution. It's just that... what I've managed to achieve (integration with Idea):

![screen shot 2016-01-25 at 11 04 13 pm](https://cloud.githubusercontent.com/assets/743666/12566055/0186a2e0-c3b8-11e5-9793-d61c99c162f1.png)

Below's the code of my ParentRunner:
```kotlin
val child1 = Description.createSuiteDescription("Child1", JUnitUniqueId.next())
val child2 = Description.createSuiteDescription("Child2", JUnitUniqueId.next())
val child3 = Description.createSuiteDescription("Child3", JUnitUniqueId.next())
val child4 = Description.createSuiteDescription("Child4", JUnitUniqueId.next())
val gchild1 = Description.createSuiteDescription("gchild1", JUnitUniqueId.next())
val gchild2 = Description.createSuiteDescription("gchild2", JUnitUniqueId.next())

override fun getDescription() = desc

val desc by lazy {
    val desc = Description.createSuiteDescription(specificationClass)!!
    desc.addChild(child1)
    desc.addChild(child2)
    desc
}

override fun run(notifier: RunNotifier?) {
    desc.addChild(child3)
    desc.addChild(child4)
    child1.addChild(gchild1)
    child2.addChild(gchild2)
    junitAction(desc, notifier!!) {
        desc.children.forEach {
            junitAction(it, notifier) {
                println("Child run")
                it.children.forEach {
                    junitAction(it, notifier) { println ("gchild run") }
                }
            }
        }
    }
}
```

Some more digging: ```JUnit4IdeaTestRunner``` is the class responsible for sending test desc tree to Idea. Is it possible to make it update the tree during exeution?

Now flow is as follows:
1. Get suite description from ParentRunner
2. Send suite tree (dk what exactly does it do - I reckon that it sends tree to be displayed by Idea). **note** that there is ```packet.addString("\n");``` call just before sending the packet - is it 
3. Run actual test

Link to source I've based on: http://grepcode.com/file/repository.grepcode.com/java/ext/com.jetbrains/intellij-idea/10.0/com/intellij/junit4/JUnit4IdeaTestRunner.java#JUnit4IdeaTestRunner.sendTree%28com.intellij.rt.execution.junit.segments.OutputObjectRegistry%2Corg.junit.runner.Description%29

Now unfortunately I'm not able to integrate runner with Idea properly just because of this limitation. Every other aspect is working (execution order, unwinding given/on/it, calling beforeOn, afterOn + I've added beforeEach and afterEach e.g. to setup/reset mocks.

Looking forward for a helping hand in this matter :)

**Comment by jskierbi** [comment]
@hastebrot generic, unwinding tree would be a very cool addition to this project - this would give possibility to limitlessly extend specs :) As for now I focused on poor man's improvements to make it work using existing defined 3 lvl hierarchy. When it works, I'll try to find some time to dig this too.

**Comment by hhariri** [comment]
@jskierbi Pinged my colleague that's responsible for the unit testing stuff in IDEA.

**Comment by akozlova** [comment]
you may fake junit runner to skip sendTree - I am not very fascinating about this as you would through away a useful functionality. If your tree is dynamic from the start to the end, then it's probably the only solution available. What if you won't add any children to the suite first and add them in `run` only?

**Comment by jskierbi** [comment]
Oki doki. Works. Results are as follows:

1. beforeOn/afterOn called in proper order (both Spek runner and Idea/JUnit integration)
2. added beforeEach/afterEach idiom, it is also called in proper order
3. Idea integration works ok, test tree is built properly. Actual tests execution is done inside ```ParentRunner#getDescription()``` with results being stored in hash map,  ```ParentRunner#run(RunNotifier)``` reports stored results for each description in tree to notifier.

My branch is here: https://github.com/jskierbi/spek/tree/feature/junit_runner_call_order_fix
Please note that this is proof of concept so code is a bit messy and it needs UTs and conflict resolution vs. mater. Will do the cleanup in following days. Comments are more than welcome, esp. reg. running tests inside ```getDescription()``` method (will it lead to some troubles?)

Excited as hell to introduce this in my project!!! :+1: 

![screen shot 2016-01-26 at 11 03 47 pm](https://cloud.githubusercontent.com/assets/743666/12596697/701efeb8-c481-11e5-8243-b3a8940db9e7.png)

Tests code:
```kotlin
beforeEach { println("    *before each") }
afterEach { println("    *after each") }

given("1") {
    println("  ->given1 expr start")
    beforeOn { println("    **beforeOn") }
    afterOn { println("    **afterOn") }
    on("1") {
        println("    ->on1 expr start")
        it ("1") {
            println("      ->on1.it1 expr")
            fail("failed it")
        }
        it ("2") { println("      ->on1.it2 expr") }
        println("    ->on1 expr end")
    }
    on("2") {
        println("    ->on2 expr start")
        it ("1") { println("      ->on2.it1 expr") }
        it ("2") { println("      ->on2.it2 expr") }
        println("    ->on2 expr end")
    }
    on("3") {
        println("    ->on3 expr start")
        it ("1") { println("      ->on3.it1 expr") }
        println("    ->on3 expr end")
    }
    println("  ->given1 expr end")
}

given("2") {
    println("  ->given 2 expr start")
    on("1") {
        println("    ->on1 expr start")
        it("1") { println("      ->on1.it1 expr") }
        it("2") { println("      ->on1.it2 expr") }
        println("    ->on1 expr end")
    }
    on("2") {
        println("    ->on2 expr start")
        it("1") { println("      ->on2.it1 expr") }
        it("2") { println("      ->on2.it2.expr") }
        println("    ->on2 expr end")
    }
    println("  ->given 2 expr end")
}
```

Logs:
```
  ->given1 expr start
  ->given1 expr end
    *before each
    **beforeOn
    ->on1 expr start
    ->on1 expr end
      ->on1.it1 expr
      ->on1.it2 expr
    **afterOn
    *after each
    *before each
    **beforeOn
    ->on2 expr start
    ->on2 expr end
      ->on2.it1 expr
      ->on2.it2 expr
    **afterOn
    *after each
    *before each
    **beforeOn
    ->on3 expr start
    ->on3 expr end
      ->on3.it1 expr
    **afterOn
    *after each
  ->given 2 expr start
  ->given 2 expr end
    *before each
    ->on1 expr start
    ->on1 expr end
      ->on1.it1 expr
      ->on1.it2 expr
    *after each
    *before each
    ->on2 expr start
    ->on2 expr end
      ->on2.it1 expr
      ->on2.it2.expr
    *after each
```

**Comment by hhariri** [comment]
@jskierbi This is fantastic! Should we test the branch already or you want to do some clean-up first?

**Comment by akozlova** [comment]
sorry, didn't check the code but running tests inside getDescription is really unexpected. Do you check that getDescription was already called and return the cach?

**Comment by jskierbi** [comment]
@hhariri let me do the cleanup first and then I'll submit a PR, hopefully this week :) I wanted to share the news as soon as I made it work

@akozlova, ```run(RunNotifier)``` bases on values produced by ```getDescription()```, traversing recursively children of root description and reporting their execution status. Basically, there is a hashmap, thats maps Description hashCode to execution result. Descriptions are created and attached to their parents while traversing execution tree, and test result is stored in hashmap for future reporting. So traversal and execution of tests is done inside ```getDesciption()``` and then another traversal over description tree is done inside ```run(RunNotifier)``` notifying each description with proper execution status. I checked impl of ```JUnit4IdeaTestRunner``` and it should work ok - ```getDescription()``` is called first, then ```run(RunNotifier)``` is called somewhere inside JUnit Runner. Also checked JUnit impl (especially ```JUnitCore``` and ```Runner```) to see whether running tests inside runner makes a change - and it seems it doesn't.

I can be wrong, though.

**Comment by lkogler** [comment]
Hey, @jskierbi I like where you're going with the PR!  But it seems to me that the behavior you've got is still not quite what I would expect, based on other libraries such as Rspec, Jasmine, etc.

I've ported your test code into Rspec to illustrate what I mean:
```ruby
require 'spec_helper'

describe 'Example' do
  before(:each) do
    puts '    *before each'
  end
  after(:each) do
    puts '    *after each'
  end

  before(:all) do
    puts '    *before all'
  end
  after(:all) do
    puts '    *after all'
  end

  describe '1' do
    puts '  ->given1 expr start'
    before do
      puts '    **beforeOn'
    end
    after do
      puts '    **afterOn'
    end

    context '1' do
      puts '    ->on1 expr start'
      it '1' do
        puts '      ->on1.it1 expr'
      end

      it '2' do
        puts '      ->on1.it2 expr'
      end
      puts '    ->on1 expr end'
    end

    context '2' do
      puts '    ->on2 expr start'
      it '1' do
        puts '      ->on2.it1 expr'
      end

      it '2' do
        puts '      ->on2.it2 expr'
      end
      puts '    ->on2 expr end'
    end

    context '3' do
      puts '    ->on3 expr start'
      it '1' do
        puts '      ->on3.it1 expr'
      end
      puts '    ->on3 expr end'
    end
    puts '  ->given1 expr end'
  end

  describe '2' do
      puts '  ->given 2 expr start'
      context '1' do
        puts '    ->on1 expr start'
        it '1' do
          puts '      ->on1.it1 expr'
        end

        it '2' do
          puts '      ->on1.it2 expr'
        end
        puts '    ->on1 expr end'
      end

      context '2' do
        puts '    ->on2 expr start'
        it '1' do
          puts '      ->on2.it1 expr'
        end

        it '2' do
          puts '      ->on2.it2.expr'
        end
        puts '    ->on2 expr end'
      end
      puts '  ->given 2 expr end'
  end
end
```

This code outputs the following:
```
  ->given1 expr start
    ->on1 expr start
    ->on1 expr end
    ->on2 expr start
    ->on2 expr end
    ->on3 expr start
    ->on3 expr end
  ->given1 expr end
  ->given 2 expr start
    ->on1 expr start
    ->on1 expr end
    ->on2 expr start
    ->on2 expr end
  ->given 2 expr end
Run options: include {:full_description=>/Example/}
    *before all
    *before each
    **beforeOn
      ->on1.it1 expr
    **afterOn
    *after each
    *before each
    **beforeOn
      ->on1.it2 expr
    **afterOn
    *after each
    *before each
    **beforeOn
      ->on2.it1 expr
    **afterOn
    *after each
    *before each
    **beforeOn
      ->on2.it2 expr
    **afterOn
    *after each
    *before each
    **beforeOn
      ->on3.it1 expr
    **afterOn
    *after each
    *before each
      ->on1.it1 expr
    *after each
    *before each
      ->on1.it2 expr
    *after each
    *before each
      ->on2.it1 expr
    *after each
    *before each
      ->on2.it2.expr
    *after each
    *after all
```

The thing to note here is that in Rspec, a regular `before` (or `before(:each)`) executes once for every individual test (`it` block).  Rspec also provides a `before(:all)`, which executes only once for the set of all tests with the context it is defined for.

**Comment by hastebrot** [comment]
Nested `context`s and `describe`s have a tricky execution order. `before(:each)` needs to execute for every `it` in the hierarchy. I experimented a bit with Mocha.js, and didn't expect the execution order.

**Comment by lkogler** [comment]
I took a stab at fixing this problem and ended up doing quite a bit of changes.  You can see my fork over at https://github.com/lkogler/spek.  Basically, I changed the DSL to behave like RSpec: you can have `describe` blocks nested to arbitrary depth, and there are `beforeEach` and `afterEach` blocks that execute before and after every `it` block. 

I know that it's quite a major change, I'd love to hear what people think before spending the time to put the finishing touches on it for a pull request.  Please let me know if this is something that sounds interesting!

**Comment by jskierbi** [comment]
@lkogler as I can see from your logs, problem is not solved, i.e. execution order is invalid for source problem of this issue. Using your solution, the result of original problem would read:
2134 
(please correct me if I'm wrong)

What I found out working on this, main issue is unwinding test execution hierarchy (tree structured) beforehand execution ```it``` blocks (leafs in the structure), which results in unexpected execution order. Please note that this is not really a problem for spek runner, but for JUnitRunner.

**Comment by hhariri** [comment]
Branch 1.0 should now solve this issue.
